File: 프로그래머스/unrated/181832. 정수를 나선형으로 배치하기/정수를 나선형으로 배치하기.py
def solution(n):
    answer = [[0]*n for _ in range(n)] 

    dr = [-1, 0, 1, 0]
    dc = [0, 1, 0, -1]

    r, c = 0, 0
    d, cnt = 1, 1

    while(n*n > cnt):
        nr = r + dr[d]
        nc = c + dc[d]
        if nr < 0 or nr >= n or nc < 0 or nc >=n or answer[nr][nc] != 0:
            d = (d + 1) % 4
            continue 
        answer[r][c] = cnt
        cnt += 1
        r = nr
        c = nc
    # if문으로 인해 마지막 원소 안 되므로 while문 끝나고 적용
    answer[r][c] = cnt
    return answer

File: 프로그래머스/0/181832. 정수를 나선형으로 배치하기/정수를 나선형으로 배치하기.java
class Solution {
    public int[][] solution(int n) {
        int[] dr = {-1, 0, 1, 0};
        int[] dc = {0, 1, 0, -1};
        int[][] answer = new int[n][n];
        
        int r = 0, c = 0;
        int d = 1, num = 1;
        
        while(n*n > num){
            int nr = r + dr[d];
            int nc = c + dc[d];

            if(nr < 0 || nc < 0 || nr >= n || nc >= n || answer[nr][nc] != 0){
                d = (d+1) % 4;
                continue;
            }
            
            answer[r][c] = num++;
            r = nr;
            c = nc;
        }
        
        answer[r][c] = num;
        return answer;
    }
}